fedimint-core/src/query.rs

```diff
@@ -163,100 +168,77 @@ impl<R: Debug + Eq + Clone> QueryStrategy<R, Vec<R>> for UnionResponsesSingle<R>
     }
 }
 
-/// Returns when `required` responses are equal, retrying after every `required`
-/// responses
-// FIXME: should be replaced by queries for specific epochs in case we cannot
-// get enough responses FIXME: for any single epoch
-pub struct EventuallyConsistent<R> {
-    responses: BTreeSet<PeerId>,
-    current: CurrentConsensus<R>,
-    required: usize,
-}
-
-impl<R> EventuallyConsistent<R> {
-    pub fn new(required: usize) -> Self {
-        Self {
-            responses: BTreeSet::new(),
-            current: CurrentConsensus::new(required),
-            required,
-        }
-    }
+/// Returns when `required` responses are equal
+pub struct CurrentConsensus<R> {
+    /// Previously received responses/errors
+    responses: BTreeMap<PeerId, R>,
+    errors: BTreeMap<PeerId, MemberError>,
+    responded_peers: BTreeSet<PeerId>,
+    threshold: usize,
+    max_evil: usize,
 }
 
-impl<R: Eq + Clone + Debug> QueryStrategy<R> for EventuallyConsistent<R> {
-    fn process(&mut self, peer: PeerId, result: api::MemberResult<R>) -> QueryStep<R> {
-        self.responses.insert(peer);
+impl<R> CurrentConsensus<R> {
+    pub fn new(total_peers: usize) -> Self {
+        let max_evil = (total_peers - 1) / 3;
```

Comment:
Having an object implementing `NumPeers` in here would also avoid this code duplication, there'd just be a `max_evil` function to call.